storage: use raft log size (in bytes) as metric for truncation

d4l3k opened the pull request:
The raft log currently truncates based of the number of entries (max 10000).
These entries can be of any size so it makes more sense to use the byte size of
the raft log as a metric for truncation.

If the raft log is greater than 64MB and there is a behind node, it will
truncate the log to the quorum committed index.

See #7065, #7201.

This PR has the changes that were discussed last Tuesday. I've also reenabled TestGetTruncatableIndexes since it seems to be fixed with these changes (verified with 4000 iterations using stress). I haven't removed the comments yet since I'm not entirely sure why.

This is a new pull request since I closed the old one while I was working on it and then force pushed which stops me from reopening it.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7438)

<!-- Reviewable:end -->

d4l3k committed:
storage: use raft log size (in bytes) as metric for truncation

The raft log currently truncates based of the number of entries (max 10000).
These entries can be of any size so it makes more sense to use the byte size of
the raft log as a metric for truncation.

If the raft log is greater than 64MB and there is a behind node, it will
truncate the log to the quorum committed index.

The in-memory size is the approximate size in bytes of the persisted raft log.
On server restart, this value is assumed to be zero to avoid costly scans of the
raft log. After the first raft log truncation it will be correct.

See #7065.

tamird commented:
Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/raft_log_queue.go, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyvV8s0uJ_hL4Orhwj:-KKyvV8s0uJ_hL4Orhwk:-1430233910) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L67)):*
> ```Go
> 
> // getTruncatableIndexes returns the total number of stale raft log entries that
> // can be truncated and the oldest index that cannot be pruned. If estimate is
> ```

remove this change

We need a comment here that describes in some detail the raft log truncation strategy.

---

*[storage/raft_log_queue.go, line 100 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyw0mDvgbAPypHHac_:-KKyw0mDvgbAPypHHaca:1307498842) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L100)):*
> ```Go
> 	oldestIndex := getMaximumMatchedIndex(raftStatus)
> 
> 	// Truncate raft logs to quorum index if the raft log is too big, or the
> ```

raftStatus.Commit is not the quorum index...

---

*[storage/raft_log_queue.go, line 160 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKywRdLyPwONO-WjZD0:-KKywRdMOsW_HZlrm9qj:-1103434340) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L160)):*
> ```Go
> }
> 
> // getMaximumMatchedIndex returns the index that has been committed by every
> ```

"the maximum index"

---

*[storage/replica.go, line 206 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyxKjyhta0IU14lXm9:-KKyxKjyhta0IU14lXmA:-1772133057) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/replica.go#L206)):*
> ```Go
> 		// Last index persisted to the raft log (not necessarily committed).
> 		lastIndex uint64
> 		// The size of the persisted raft log.
> ```

"The size in bytes"

---

*[storage/replica_raftstorage.go, line 457 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyx7HLRdzYS6O-1QOx:-KKyx7HMl1hLdqpaBc5B:-632527180) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/replica_raftstorage.go#L457)):*
> ```Go
> // of r.lastIndex and returns a new value. We do this rather than
> // modifying r.lastIndex directly because this modification needs to
> // be atomic with the commit of the batch.
> ```

needs an update

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks broke.

---

*[storage/raft_log_queue.go, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyvV8s0uJ_hL4Orhwj:-KKz0KeBDWbZHMqTQmiW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L67)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove this change
> 
> We need a comment here that describes in some detail the raft log truncation strategy.
</details>
Done.

---

*[storage/raft_log_queue.go, line 100 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyw0mDvgbAPypHHac_:-KKz-56EiLSqyD5jpH-b:-438543145) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L100)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> raftStatus.Commit is not the quorum index...
</details>
How so? Isn't it the latest committed index that the current node knows about? Hence the quorum committed index?

---

*[storage/raft_log_queue.go, line 160 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKywRdLyPwONO-WjZD0:-KKyzl4al7ubY1mN5o-C:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L160)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "the maximum index"
</details>
Done.

---

*[storage/replica.go, line 206 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyxKjyhta0IU14lXm9:-KKyztJPpAU0ouPRiCSX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/replica.go#L206)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "The size in bytes"
</details>
Done.

---

*[storage/replica_raftstorage.go, line 457 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyx7HLRdzYS6O-1QOx:-KKz-nUL5htWlfvQa_9N:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/replica_raftstorage.go#L457)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> needs an update
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/raft_log_queue.go, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyvV8s0uJ_hL4Orhwj:-KKz19arxz7Um4AqGSxs:435017343) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L67)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Done.
</details>
This isn't detailed enough. An uninitiated reader will not understand what this really does from this description.

---

*[storage/raft_log_queue.go, line 101 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKz1P48biFUgYRFopcD:-KKz1P48biFUgYRFopcE:-101564440) ([raw file](https://github.com/cockroachdb/cockroach/blob/11bd7ddafd6824f4fee04cebab8402b324eb13c3/storage/raft_log_queue.go#L101)):*
> ```Go
> 
> 	// Truncate raft logs to the quorum committed index if the raft log is too
> 	// big, or the oldest node is too far behind.
> ```

can you refine "too far behind"? probably "or there exists a node which can only be caught up by snapshot", but even that description is anemic. I think this comment should be reworked completely to make the consequences of this decision clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/raft_log_queue.go, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyvV8s0uJ_hL4Orhwj:-KKz6dJz86Np3fx6OGTD:91859267) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L67)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> This isn't detailed enough. An uninitiated reader will not understand what this really does from this description.
</details>
But then they won't have to suffer like I did! 

Better?

---

*[storage/raft_log_queue.go, line 101 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKz1P48biFUgYRFopcD:-KKz6iKjruXo8loc_631:-698231926) ([raw file](https://github.com/cockroachdb/cockroach/blob/11bd7ddafd6824f4fee04cebab8402b324eb13c3/storage/raft_log_queue.go#L101)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you refine "too far behind"? probably "or there exists a node which can only be caught up by snapshot", but even that description is anemic. I think this comment should be reworked completely to make the consequences of this decision clear.
</details>
Better?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/raft_log_queue.go, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKyvV8s0uJ_hL4Orhwj:-KKz7xSF5Ll0KzvC-wV1:-843633951) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L67)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> But then they won't have to suffer like I did! 
> 
> Better?
</details>
Still find this hard to read, but OK.

---

*[storage/raft_log_queue.go, line 102 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKz8BH_JIii0Cv6j0ZV:-KKz8BHaFyF04z6WUlzO:-457388199) ([raw file](https://github.com/cockroachdb/cockroach/blob/a64592b896080229fa0e48f6d8182bdfc2ebe400/storage/raft_log_queue.go#L102)):*
> ```Go
> 
> 	// Always truncate the oldest raft log entry which has been committed by all replicas.
> 	oldestIndex := getMaximumMatchedIndex(raftStatus)
> ```

this is misnamed, right? it's the latest (and more accurately maximum) index, not the oldest. This misnaming also causes the below comment to be difficult to parse.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/raft_log_queue.go, line 102 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KKz8BH_JIii0Cv6j0ZV:-KKzAZTKHoiLDQqSn0Ph:1459638738) ([raw file](https://github.com/cockroachdb/cockroach/blob/a64592b896080229fa0e48f6d8182bdfc2ebe400/storage/raft_log_queue.go#L102)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is misnamed, right? it's the latest (and more accurately maximum) index, not the oldest. This misnaming also causes the below comment to be difficult to parse.
</details>
It's the index that we can truncate to, which could be thought of as the "oldest index" we keep. Renamed and clarified the below comment. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. Mind adding in the commit message that we keep track only in-memory, and that that's sufficient (can be the same text as requested in one of my comments below)

---

Reviewed 13 of 13 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/raft_log_queue.go, line 169 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KL1vo31YKxmkSNlVHuk:-KL1vo32CxT90yE4yFDP:-1610104564) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L169)):*
> ```Go
> 		}
> 	}
> 	return smallestMatch
> ```

Whatever you're returning here must never be `> raftStatus.Commit`. Pretty sure that's not violated (because that means something being in all logs, but not having been committed, and Raft would have updated the commit before it got here), but let's be paranoid and enforce `smallestMatch <= Commit`.

---

*[storage/raft_log_queue.go, line 73 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KL1yiaX70IRpM0tqTPK:-KL1yiaX70IRpM0tqTPL:1155570606) ([raw file](https://github.com/cockroachdb/cockroach/blob/a64592b896080229fa0e48f6d8182bdfc2ebe400/storage/raft_log_queue.go#L73)):*
> via the raft log (due to a previous truncation),

---

*[storage/raft_log_queue_test.go, line 63 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KL1wOzH4Vuxck_xh1AD:-KL1wOzIZG0kMNEwslGv:610327397) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue_test.go#L63)):*
> ```Go
> func TestGetTruncatableIndexes(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	//t.Skip("TODO(bram): #7056")
> ```

Remove.

---

*[storage/replica.go, line 206 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KL1xmAUVG3jtMABsGD3:-KL1xmAUVG3jtMABsGD4:-1679284978) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/replica.go#L206)):*
> ```Go
> 		// Last index persisted to the raft log (not necessarily committed).
> 		lastIndex uint64
> 		// The size of the persisted raft log.
> ```

More words. Someone who wasn't in the room when the semantics were discussed will have no idea how this field behaves.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-:-KL1zN31-bZEFmBZlDh2:2002560646)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Added.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/raft_log_queue.go, line 169 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KL1vo31YKxmkSNlVHuk:-KLT308fM-Nt4hN_u6Uq:447191349) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue.go#L169)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Whatever you're returning here must never be `> raftStatus.Commit`. Pretty sure that's not violated (because that means something being in all logs, but not having been committed, and Raft would have updated the commit before it got here), but let's be paranoid and enforce `smallestMatch <= Commit`.
</details>
Done. Changed the upper bound from `math.MaxUint64` to `raftStatus.Commit`.

---

*[storage/raft_log_queue.go, line 73 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KL1yiaX70IRpM0tqTPK:-KLT3F4FJD2Z6odqL1rk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a64592b896080229fa0e48f6d8182bdfc2ebe400/storage/raft_log_queue.go#L73)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> > via the raft log (due to a previous truncation),
</details>
Done.

---

*[storage/raft_log_queue_test.go, line 63 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KL1wOzH4Vuxck_xh1AD:-KLT3h92dXJFQ88_eqp9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/raft_log_queue_test.go#L63)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Remove.
</details>
Done.

---

*[storage/replica.go, line 206 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KL1xmAUVG3jtMABsGD3:-KLT6by1g_XEa6H527Or:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a024e546b665803a9fa07efe18a699a4f106ed8f/storage/replica.go#L206)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> More words. Someone who wasn't in the room when the semantics were discussed will have no idea how this field behaves.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-:-KLT6aTygsQ8lmhOU6Fp:1610000531)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 11 of 11 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 11 of 11 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/raft_log_queue.go, line 105 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLlwcVvS9VmCyE0zErT:-KLlwcVwNoGTavaQz2BM:-1379398394) ([raw file](https://github.com/cockroachdb/cockroach/blob/91affb4ab778e6d2c81dca6610ee383a094bbc7c/storage/raft_log_queue.go#L105)):*
> ```Go
> 
> 	// Truncate raft logs to the quorum committed index if the raft log is too
> 	// big. If the most behind node's committed index is before the first raft log
> ```

This is fine for three-replica configurations, but is suboptimal for five replicas. In a five-replica configuration with one node far behind and one node slightly behind, truncating to the commit index will require the node that is only slightly behind to apply a snapshot too. I think we should aim to only cut off one node per truncation: iterate over `raftStatus.Progress` to find the smallest match index which is greater than firstIndex and use that for the truncation.

It's possible that we're still over the limit after doing this (if there are two far-behind nodes), but the next iteration of the queue will truncate the log further. 

---

*[storage/replica.go, line 210 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLm0P8DZZIUzP3o8GMW:-KLm0P8DZZIUzP3o8GMX:-1152318419) ([raw file](https://github.com/cockroachdb/cockroach/blob/91affb4ab778e6d2c81dca6610ee383a094bbc7c/storage/replica.go#L210)):*
> ```Go
> 		// raftLogSize is the approximate size in bytes of the persisted raft log.
> 		// On server restart, this value is assumed to be zero to avoid costly scans
> 		// of the raft log. After the first raft log truncation it will be correct.
> ```

A single truncation won't necessarily make this correct (see the check for negative values in TruncateLog). Instead, it is correct when all the remaining entries have been written during the lifetime of this process. 

---

*[storage/storagebase/state.proto, line 62 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLlyAJfbW8qzIDET4ER:-KLlyAJfbW8qzIDET4ES:-475456351) ([raw file](https://github.com/cockroachdb/cockroach/blob/91affb4ab778e6d2c81dca6610ee383a094bbc7c/storage/storagebase/state.proto#L62)):*
> ```protobuf
>   uint64 num_pending = 3;
>   util.hlc.Timestamp last_verification = 4 [(gogoproto.nullable) = false];
>   int64 raft_log_size = 5;
> ```

Comment that this may be inaccurate on nodes which have recently restarted. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-:-KLm14evaaTBINlQiKoO:1417088721)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/raft_log_queue.go, line 105 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLlwcVvS9VmCyE0zErT:-KLw8XAPHc_wUjBrDHL8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/91affb4ab778e6d2c81dca6610ee383a094bbc7c/storage/raft_log_queue.go#L105)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is fine for three-replica configurations, but is suboptimal for five replicas. In a five-replica configuration with one node far behind and one node slightly behind, truncating to the commit index will require the node that is only slightly behind to apply a snapshot too. I think we should aim to only cut off one node per truncation: iterate over `raftStatus.Progress` to find the smallest match index which is greater than firstIndex and use that for the truncation.
> 
> It's possible that we're still over the limit after doing this (if there are two far-behind nodes), but the next iteration of the queue will truncate the log further. 
</details>
Done.

---

*[storage/replica.go, line 210 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLm0P8DZZIUzP3o8GMW:-KLvyeeosm2ZviWggwbC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/91affb4ab778e6d2c81dca6610ee383a094bbc7c/storage/replica.go#L210)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> A single truncation won't necessarily make this correct (see the check for negative values in TruncateLog). Instead, it is correct when all the remaining entries have been written during the lifetime of this process. 
</details>
Done.

---

*[storage/storagebase/state.proto, line 62 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLlyAJfbW8qzIDET4ER:-KLvye6I8kv-dX9GCjHB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/91affb4ab778e6d2c81dca6610ee383a094bbc7c/storage/storagebase/state.proto#L62)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Comment that this may be inaccurate on nodes which have recently restarted. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 14 files at r7.
Review status: 13 of 14 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[storage/raft_log_queue.go, line 67 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLweabp0H3Irh1Nf_BI:-KLweabp0H3Irh1Nf_BJ:87645838) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L67)):*
> ```Go
> 
> // getTruncatableIndexes returns the number of truncatable indexes and the
> // oldest index that cannot be pruned for the replica.
> ```

s/pruned/truncated/ unless we use pruned elsewhere.

---

*[storage/raft_log_queue.go, line 99 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLweehByqktgncslMXj:-KLweehByqktgncslMXk:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L99)):*
> ```Go
> }
> 
> // computeTruncatableIndex returns the oldest index that cannot be pruned. If
> ```

ditto

---

*[storage/raft_log_queue.go, line 106 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwem1rzapDpB0lj3_J:-KLwem1rzapDpB0lj3_K:268237707) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L106)):*
> ```Go
> // bytes value (typically 64MB). When there are no nodes behind, or we can't
> // catch them up via the raft log (due to a previous truncation) this returns
> // the quorum committed index.
> ```

no longer true

---

*[storage/raft_log_queue.go, line 110 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwfB5vckvBprJ0-ntY:-KLwfB5vckvBprJ0-ntZ:1265641811) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L110)):*
> ```Go
> 	truncatableIndex := raftStatus.Commit
> 	behindIndexes := getBehindIndexes(raftStatus)
> 	for _, i := range behindIndexes {
> ```

`i` is typically used as a loop variable, can you name this the singular of the slice being iterated?

---

*[storage/raft_log_queue.go, line 120 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwfRRM2GcoxyYyDsb7:-KLwfRRM2GcoxyYyDsb8:911301719) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L120)):*
> ```Go
> 		// quorum committed index. This allows for multiple nodes to be behind and not
> 		// give up on the more recently behind node.
> 		if raftLogSize > targetSize && i <= firstIndex {
> ```

why can't this be:
```
if raftLogSize <= targetSize {
  truncatableIndex = i
  break
}
```

---

*[storage/replica.go, line 210 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLm0P8DZZIUzP3o8GMW:-KLweIFqGz4NU9k9R-zs:2092133899) ([raw file](https://github.com/cockroachdb/cockroach/blob/91affb4ab778e6d2c81dca6610ee383a094bbc7c/storage/replica.go#L210)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Done.
</details>
pedantic nit: "This will be correct when all log entries predating this process have been truncated."

---

*[storage/storagebase/state.proto, line 64 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLweUa-w9ylEPr-ON6c:-KLweUa-w9ylEPr-ON6d:-1338051834) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/storagebase/state.proto#L64)):*
> ```protobuf
>   uint64 num_dropped = 5;
>   // raft_log_size may be initially inaccurate after a server restart.
>   // See storage.Replica.mu.raftLogSize
> ```

period

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 10 of 14 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[storage/raft_log_queue.go, line 67 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLweabp0H3Irh1Nf_BI:-KLwi-IuTMSw80YcsHxr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L67)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/pruned/truncated/ unless we use pruned elsewhere.
</details>
Done.

---

*[storage/raft_log_queue.go, line 99 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLweehByqktgncslMXj:-KLwi-dpiUHAk0GsGiUb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L99)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[storage/raft_log_queue.go, line 106 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwem1rzapDpB0lj3_J:-KLwoOCNm6c9iaV2ng52:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L106)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> no longer true
</details>
Done.

---

*[storage/raft_log_queue.go, line 110 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwfB5vckvBprJ0-ntY:-KLwjMawKq2v7bQB841W:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L110)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `i` is typically used as a loop variable, can you name this the singular of the slice being iterated?
</details>
Done.

---

*[storage/raft_log_queue.go, line 120 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwfRRM2GcoxyYyDsb7:-KLwoR95db7sv26IwxUf:986514180) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/raft_log_queue.go#L120)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why can't this be:
> ```
> if raftLogSize <= targetSize {
>   truncatableIndex = i
>   break
> }
> ```
</details>
I simplified the logic a bit.

When the `raftLogSize <= targetSize` we want to keep the log as is when it's already been truncated to the furthest behind node.

Otherwise, we want to truncate to the next behind node.

---

*[storage/replica.go, line 210 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLm0P8DZZIUzP3o8GMW:-KLwl6ZL-qGhLbCRpRHD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/91affb4ab778e6d2c81dca6610ee383a094bbc7c/storage/replica.go#L210)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> pedantic nit: "This will be correct when all log entries predating this process have been truncated."
</details>
Done.

---

*[storage/storagebase/state.proto, line 64 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLweUa-w9ylEPr-ON6c:-KLwjgEX2XT9PD8FPHEg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa9631ad539946c6dcc1bf45aa4c3849a971d041/storage/storagebase/state.proto#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> period
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r8.
Review status: 13 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/raft_log_queue.go, line 112 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwxNSUZwDPEFFHrEgI:-KLwxNSUZwDPEFFHrEgJ:-1016107945) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0c8b9639568b989b5a155f71affabe6a8460c/storage/raft_log_queue.go#L112)):*
> ```Go
> 	behindIndexes := getBehindIndexes(raftStatus)
> 	for _, behindIndex := range behindIndexes {
> 		// If the behind node's committed index is before the first raft log
> ```

this comment is now hard to read given the code below is testing for the inverse condition.

---

*[storage/raft_log_queue_test.go, line 32 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwxDT6mW0l1QHd2hjV:-KLwxDT7nIMBUkWcKQ2j:-2080343203) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0c8b9639568b989b5a155f71affabe6a8460c/storage/raft_log_queue_test.go#L32)):*
> ```Go
> )
> 
> // TestGetBehindIndexes verifies that the correc
> ```

missing the rest of the sentence

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 11 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/raft_log_queue.go, line 112 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwxNSUZwDPEFFHrEgI:-KM-x_lwWUwnaHHFJAwt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0c8b9639568b989b5a155f71affabe6a8460c/storage/raft_log_queue.go#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment is now hard to read given the code below is testing for the inverse condition.
</details>
Done.

---

*[storage/raft_log_queue_test.go, line 32 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-KLwxDT6mW0l1QHd2hjV:-KM-xadxI1divtZHAtu8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ec0c8b9639568b989b5a155f71affabe6a8460c/storage/raft_log_queue_test.go#L32)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> missing the rest of the sentence
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Since the LGTMs are stale. Everyone happy with me merging this?

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 14 files at r7, 1 of 3 files at r8, 3 of 3 files at r9, 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7438#-:-KM67q4dD1pG3vu--JX7:1417088721)*
<!-- Sent from Reviewable.io -->
